woocommerce/woocommerce
# Track first and last 20 orders to Weekly Tracker Snapshot

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR will include the tracking date for the first and last 20 orders in the Weekly Tracker Snapshot. The following fields will be tracked for each of these orders:
- Order Rank
- Date created
- Date updated
- Currency
- Amount
- Payment method
- Payment method title
- Refund Amount
- Recorded sales
- WooCommerce version

The tracking will support both HPOS as well as legacy order storage in the post tables. In order to speed up the process of fetching this data we fetch it in several queries to prevent multiple joins of potentially large tables.

In addition to the order snapshot we also include the total time to generate the snapshot data, this can be used in the future to monitor if generating the snapshot data takes an excessive amount of time on some sites.

Part of the project thread pffQ75-V7-p2

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

The easiest way to test the changes is to use a code snippet which will display the snapshot data directly within the admin screen. The following snippet can be used for this:
```php
add_action(
	'admin_init',
	function () {
		if ( ! empty( $_GET['wctracker'] ) ) {
			$tracks = WC_Tracker::get_tracking_data();
			echo '<pre>' . json_encode( $tracks, JSON_PRETTY_PRINT ) . '</pre>';
			exit;
		}
	}
);
```

We can then visit any admin page and set the `wctracker` parameter to view the data, for example with a URL like this:
`https://domain.test/wp-admin/?wctracker=1`

1. Generate a set of orders on the site (initially less than 20)
2. Refund some orders to confirm refund data will be included
3. Enable the code snippet and visit the URL to view the tracking data `https://domain.test/wp-admin/?wctracker=1`
4. Confirm that the `order_snapshot` and `snapshot_generation_time` is populated, which could look something like below:
```
{
    ...
    "order_snapshot": {
        "first_20_orders": {
            "523": {
                "id": "523",
                "order_rank": "1",
                "order_created_at": "2023-01-11 13:00:00",
                "order_updated_at": "2024-07-04 07:44:59",
                "currency": "EUR",
                "total_amount": "25817.54000000",
                "payment_method": "stripe",
                "payment_method_title": "Stripe",
                "woocommerce_version": "9.2.0",
                "recorded_sales": "yes"
            },
            "581": {
                "id": "581",
                "order_rank": "2",
                "order_created_at": "2023-01-12 22:00:00",
                "order_updated_at": "2024-07-04 07:47:46",
                "currency": "EUR",
                "total_amount": "20808.71000000",
                "payment_method": "cod",
                "payment_method_title": "Cash on Delivery",
                "woocommerce_version": "9.2.0",
                "recorded_sales": "yes"
            },
            ...
        "last_20_orders": {
            "724": {
                "id": "724",
                "order_rank": "183",
                "order_created_at": "2024-10-26 07:00:00",
                "order_updated_at": "2024-10-16 12:58:58",
                "currency": "EUR",
                "total_amount": "8151.81000000",
                "payment_method": "stripe",
                "payment_method_title": "Stripe",
                "woocommerce_version": "9.5.0",
                "recorded_sales": "yes",
                "refund_amount": "-528.27000000"
            },
            "725": {
                "id": "725",
                "order_rank": "181",
                "order_created_at": "2024-10-25 03:00:00",
                "order_updated_at": "2024-08-26 11:22:56",
                "currency": "EUR",
                "total_amount": "7098.33000000",
                "payment_method": "stripe",
                "payment_method_title": "Stripe",
                "woocommerce_version": "9.3.0",
                "recorded_sales": "yes"
            },
    ...
    "snapshot_generation_time": 0.028749942779541016
}
```
5. Generation additional order data for more than 40 orders
6. Confirm that the order data is still tracked and the first and last orders are a different set
7. Go to WooCommerce > Settings > Advanced > Features and toggle between HPOS and legacy order storage
8. Check the tracking snapshot data again and confirm it contains the same tracking details

### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Track first and last 20 orders to Weekly Tracker Snapshot
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
